### PR TITLE
Adding support for SmartThings set target temperature async.

### PR DESCRIPTION
**Pull request by EllenPorter**

### CODE REVIEW

**Comment by EllenPorter** [comment]
Need to update mock response and remove old resourceSchemaToProviderSchema.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 1]
```diff
@@ -137,20 +139,20 @@ function providerSchemaToPlatformSchema(providerSchema, expand) {
 
     var hvacMode = createResource('oic.r.mode', 'oic.if.a', 'hvacMode', expand, {
         supportedModes: ['coolOnly', 'heatOnly', 'auto', 'off'],
-        modes: [deviceHvacModeToTranslatorHvacMode(providerSchema['attributes'].thermostatMode)]
+        modes: [deviceHvacModeToTranslatorHvacMode(attributes.thermostatMode)]
     });
 
     var hasFan = createResource('oic.r.sensor', 'oic.if.s', 'hasFan', expand, {
-        value: providerSchema['attributes'].thermostatFanMode !== undefined && providerSchema['attributes'].thermostatFanMode !== null
+        value: attributes.thermostatFanMode !== undefined && attributes.thermostatFanMode !== null
```
Comment (rgruian) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: I think this also does the same thing and is simpler:
value: ((typeof attributes.thermostatFanMode) !== "undefined")

Comment (muralidhars) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: @rgruian - typeof null is object. So you need to check for null and undefined separately

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 2]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+    result['coolingSetpoint'] = newTargetTemperatureHigh;
+    result['heatingSetpoint'] = newTargetTemperatureLow;
+
+    return result;
+}
+
+function getMaxHeatingSetpoint(attributes) {
+    if (getUnitSafe(attributes) === 'f') {
+        return getIntSafe(attributes.maxHeatingSetpoint, 50);
+    } else {
```
Comment (rgruian) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: minor: else not needed

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 3]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+        return getIntSafe(attributes.maxHeatingSetpoint, 9);
+    }
+}
+
+function getMinHeatingSetpoint(attributes) {
+    if (getUnitSafe(attributes) === 'f') {
+        return getIntSafe(attributes.minHeatingSetpoint, 50);
+    } else {
```
Comment (rgruian) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: minor: else not needed

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 4]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+        return getIntSafe(attributes.minHeatingSetpoint, 9);
+    }
+}
+
+function getMaxCoolingSetpoint(attributes) {
+    if (getUnitSafe(attributes) === 'f') {
+        return getIntSafe(attributes.maxCoolingSetpoint, 90);
+    } else {
```
Comment (rgruian) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: minor: else not needed

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 5]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+        return getIntSafe(attributes.maxCoolingSetpoint, 32);
+    }
+}
+
+function getMinCoolingSetpoint(attributes) {
+    if (getUnitSafe(attributes) === 'f') {
+        return getIntSafe(attributes.minCoolingSetpoint, 90);
+    } else {
```
Comment (rgruian) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: minor: else not needed

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 6]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+        return getIntSafe(attributes.minCoolingSetpoint, 32);
+    }
+}
+
+/**
+ * SmartThings attributes can be strings, check for existance and return as int
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: typo. existence

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 7]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+ * @param {*} i 
+ */
+function getIntSafe(i) {
+    return i !== undefined && i != null ? parseInt(i) : 0;
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: return getIntSafe(i, 0);

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 8]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+                return 'c';
+            }
+        }
+        throw new OpenT2TError(440, "Temperature outside supported range (" + value + ")");
+    }
+}
+
+/**
+ * Given a target temperature, compute a target 
+ * low and target high centerd around it.
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: typo. centered

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 9]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+ * @param {*} resourceSchema 
+ * @param {*} stateReader 
+ */
+function getTargetTemperatureRange(resourceSchema, attributes) {
+    var result = {};
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: do this at line 355 i.e. closer to where it's needed

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 10]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+    var range = targetTemperatureHigh - targetTemperatureLow;
+    var halfRange = range / 2;
+    var newTargetTemperatureHigh = value + halfRange;
+    var newTargetTemperatureLow = value - halfRange;
+
+    // If either value is outside min/max range, adjust accordingly
+    if (newTargetTemperatureLow < minTemperatureLow) {
+        newTargetTemperatureLow = minTemperatureLow;
+        newTargetTemperatureHigh = minTemperatureLow + range;
+    }
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: did you miss an 'else' here?

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 11]
```diff
@@ -238,9 +240,184 @@ function validateResourceGet(resourceId) {
...
+        newTargetTemperatureLow = maxTemperatureLow;
+        newTargetTemperatureHigh = maxTemperatureLow + range;
+    }
+
+    if (newTargetTemperatureHigh > maxTemperatureHigh) {
+        newTargetTemperatureHigh = maxTemperatureHigh;
+        newTargetTemperatureLow = Math.max(minTemperatureLow, maxTemperatureHigh - range);
+    }
+
+    if (newTargetTemperatureHigh < minTemperatureHigh) {
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: same here. else if?

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 12]
```diff
@@ -285,20 +462,28 @@ class Translator {
...
-        if (di === generateGUID( this.controlId + 'opent2t.d.thermostat') )
-        {
-            var putPayload = resourceSchemaToProviderSchema(resourceId, payload);
-
-            return this.smartThingsHub.putDeviceDetailsAsync(this.endpointUri, this.controlId, putPayload)
-                .then((response) => {
-                    var schema = providerSchemaToPlatformSchema(response, true);
-                    return findResource(schema, di, resourceId);
-                });
+        if (di === generateGUID(this.controlId + 'opent2t.d.thermostat')) {
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: suggestion: invert the condition to reduce nesting

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js [hunk 13]
```diff
@@ -410,6 +595,52 @@ class Translator {
         subscriptionInfo.endpointUri = this.endpointUri;
         return this.smartThingsHub._unsubscribe(subscriptionInfo);
     }
+
+    _resourceSchemaToProviderSchemaAsync(resourceId, resourceSchema) {
+
+        // build the object with desired state
+        var result = {};
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: by doing Object.create(null) you won't get any of the basic functions attached to the object whereas with {} you do. depends on your use-case

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 1 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: value: attributes.thermostatFanMode ? true : false

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 2 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 3 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 4 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 5 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 6 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 7 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 8 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 9 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 10 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Keeping it consistent with the other translators.  if/else could be more efficient.  Will add to clean up.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 11 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Same.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 12 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Adding to code clean up task.

File: org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js (hunk 13 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.smartthings.thermostat/js/thingTranslator.js]: Keeping consistent with other translators, will add to code clean up task.

**Pull request merged**